firmware/shared/code/I2C.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * I2C.c
+ *
+ *  Created on: Oct 10, 2020
+ *      Author: Shuhang Tan
+ *      Reference: https://github.com/uvic-auvic/Firmware-Motor_Controller_2.0/blob/master/src/I2C.c
+ *                 (I2C code for motor controller 2)
+ */
+
+#include "I2C.h"
+#include <string.h>
+#include <stdlib.h>
+#include "stm32f4xx.h"
+//#include "FreeRTOS.h"
+//#include "task.h"
+#include "LED.h"
+
+// I2C SDA and SCL pins
+#define _I2C_SDA_GPIO  (GPIO_Pin_9) // PC9
+#define _I2C_SCL_GPIO  (GPIO_Pin_8) // PA8
+
+// I2C sendBuffer size
+#define _I2C_BUFFER_SIZE    (10U)
+
+typedef struct
+{
+	uint8_t Buffer[_I2C_BUFFER_SIZE];
+} I2C_data_S;
+
+// Adding an idle state because the BUSY bit in SR2 is hard to deal with
+typedef enum{
+	I2C_STATE_SEND,    // Write
+	I2C_STATE_RECEIVE, // Read
+	I2C_STATE_IDLE,    // Idle
+}I2C_state_S;
+
+// Global variables and their initialization
+static I2C_data_S I2C_send_data;
+static I2C_data_S I2C_receive_data;
+static I2C_state_S I2C_state = I2C_STATE_IDLE;
+
+uint8_t slave_address = 0x0;
+
+// Length of the data to be sent or received (in bytes)
+volatile uint8_t data_length = 0;
+
+// Data counter, 0 when a process finished or not started
+volatile uint8_t data_count = 0;
+
+// Pointer to the buffers

```

Comment:
Since these two are only used in the ISR, you might be able to bring down global variable count by declaring them as static within there

Reply:
data_length is used in both send and receive function. data_count is not useful in the code and has been deleted.